src/groupeddataframe/splitapplycombine.jl

```diff
@@ -499,25 +512,40 @@ wrap(x::AbstractMatrix) =
     NamedTuple{Tuple(gennames(size(x, 2)))}(Tuple(view(x, :, i) for i in 1:size(x, 2)))
 wrap(x::Any) = (x1=x,)
 
-wrap_table(x::Any) =
+wrap_table(x::Any, ::Val) =
     throw(ArgumentError("return value must not change its kind " *
                         "(single row or variable number of rows) across groups"))
-wrap_table(x::Union{NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}},
-                    AbstractDataFrame,
-                    AbstractVector,
-                    AbstractMatrix}) = wrap(x)
+function wrap_table(x::Union{NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}},
+                             AbstractDataFrame, AbstractMatrix}, ::Val{T}) where T
+    T || throw(ArgumentError("mixing single and multiple column values is not allowed"))
+    wrap(x)
+end
+
+function wrap_table(x::AbstractVector, ::Val{T}) where T
+    T && throw(ArgumentError("mixing single and multiple column values is not allowed"))
+    wrap(x)
+end
+
+function wrap_row(x::Any, ::Val{T}) where T
+    # NamedTuple is not possible in this branch
+    if (x isa DataFrameRow) ⊻ T
```

Comment:
Shouldn't this be the opposite? If `x` is a row, then `wasmulti` must be `false`, otherwise an error must be thrown.

Reply:
this is exactly what exclusive or `⊻` says - if both arguments have the same logical value do not throw an error.

Reply:
Mmm, wait. Isn't `x isa DataFrameRow` and `wasmulti == false` a valid situation? I thought the idea was that if `wasmulti == true` we don't want to accept single rows?

Reply:
Actually the xor was a red herring: I just confused multi-column and multi-row. Though maybe this would justify renaming `wasmulti` to e.g. `multicol`?

Reply:
OK - I rename it to `firstmulticol` now (as I have posted below we have "two dimenstions" of "multi" as you rightly note here :smile:). I will add `first` to highlight that we use the return type obtained for the first group

Reply:
pushed